src/zap/zapinfo.cpp

```diff
@@ -2083,6 +2102,59 @@ void ZapInfo::GetProfilingHandle(BOOL                      *pbHookFunction,
     *pbIndirectedHandles = TRUE;
 }
 
+DWORD FilterMethodAttribsForIsSupported(DWORD attribs, CORINFO_METHOD_HANDLE ftn, ICorDynamicInfo* pJitInfo)
+{
+    if (attribs & CORINFO_FLG_JIT_INTRINSIC)
+    {
+        // Do not report the get_IsSupported method as an intrinsic. This will turn the call into a regular
+        // call. We also make sure none of the hardware intrinsic method bodies get pregenerated in crossgen
+        // (see ZapInfo::CompileMethod) but get JITted instead. The JITted method will have the correct
+        // answer for the CPU the code is running on.
+
+        const char* namespaceName;
+        const char* className;
+        const char* enclosingClassName;
+        const char* methodName = pJitInfo->getMethodNameFromMetadata(ftn, &className, &namespaceName, &enclosingClassName);
+
+        // If it's not the IsSupported method, we're done
+        if (strcmp(methodName, "get_IsSupported") != 0)
+            return attribs;
+
+        bool isX86intrinsic = strcmp(namespaceName, "System.Runtime.Intrinsics.X86") == 0;
+
+#if defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+        // If it's IsSupported on Sse/Sse2, we can expand unconditionally since this is reliably
+        // available everywhere.
+        if (isX86intrinsic
+            && (
+                strcmp(className, "Sse") == 0 || strcmp(className, "Sse2") == 0
+                || (
+                    strcmp(className, "X64") == 0
+                    && (
+                        strcmp(enclosingClassName, "Sse") == 0 || strcmp(enclosingClassName, "Sse2") == 0
+                        )
+                    )
+                )
+            )
+            return attribs;
+#endif // defined(_TARGET_X86_) || defined(_TARGET_AMD64_)
+
+        // If it's another get_IsSupported method on some other ISA class, do not report as intrinsic
+        if (isX86intrinsic
+            || strcmp(namespaceName, "System.Runtime.Intrinsics.Arm.Arm64") == 0
```

Comment:
Do we not want to unconditionally expand the ARM `IsSupported` intrinsics as `false` on x86` (and vice-versa on ARM)?

Reply:
Yes, we should. So, I think the summary is:
- IsSupported should be treated as an intrinsic for baseline intrinsics (Sse and Sse2 for X86)
- It should remain a call for non-VEX-encoded intrinsics on X86
- For VEX-encoded intrinsics, it should disable pre-jitting.
- For non-platform intrinsics (Arm64 when on x86) it should be treated as an intrinsic (and will return `false`)

Reply:
This is how it works. Note that because of how CoreLib is structured:

https://github.com/dotnet/coreclr/blob/f70085f37f3c49618087f852e44165954b3ed8d6/src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems#L1333-L1346

ARM64 intrinsics on non-ARM64 systems are not actually intrinsics (it's a type like any other with no `[Intrinsic]` marking that just returns false/throws). So some of these paths are always unreachable (I had it under ifdef originally, but removed the ifdef per request here). `CORINFO_FLG_JIT_INTRINSIC` is not set outside the applicable architecture.

Reply:
The IsSupported Intrinsics are special and are handled as intrinsic even in the PlatformNotSupported.cs case.

This was done in https://github.com/dotnet/coreclr/pull/23751 to ensure that they are properly inlined and treated as constants, even in complex methods.